<p style="text-align: justify; text-justify: inter-word;">
5,424 unique TF DNA-binding sites (TFBSs) motifs are collected in the form of PFM and mined.
<br><br>
<h4><b>The motifs are collected from the databases:</b></h4>
<ul>
<li>Jaspar<sup>[1]</sup> - JASPAR_v2014, JASPAR_v2016, JASPAR_CORE</li>
<li>hPDI<sup>[2]</sup> </li>
<li>SwissRegulon<sup>[3]</sup></li>
<li>HOCOMOCO<sup>[4]</sup> - HOCOMOCO v10</li>
<li>TRANSFAC Professional<sup>[5]</sup> - TRANSFAC v2017.1</li>
</ul>

The annotation file for all the TFBS motifs is available to download in the <i>Download page</i>.
<br><br>
<small>[1] Khan, A., Fornes, O., Stigliani, A., et al. (2018) JASPAR 2018: update of the open-access database of transcription factor binding profiles and its web framework. Nucleic Acids Res., 46, D1284.<br>
[2] Xie, Z., Hu, S., Blackshaw, S., et al. (2010) hPDI: a database of experimental human protein-DNA interactions. Bioinformatics, 26, 287–289.<br>
[3] Pachkov, M., Balwierz, P. J., Arnold, P., et al. (2013) SwissRegulon, a database of genome-wide annotations of regulatory sites: recent updates. Nucleic Acids Res., 41, D214-220.<br>
[4] Kulakovskiy, I. V., Vorontsov, I. E., Yevshin, I. S., et al. (2018) HOCOMOCO: towards a complete collection of transcription factor binding models for human and mouse via large-scale ChIP-Seq analysis. Nucleic Acids Res., 46, D252–D259.<br>
[5] Matys, V., Kel-Margoulis, O. V., Fricke, E., et al. (2006) TRANSFAC and its module TRANSCompel: transcriptional gene regulation in eukaryotes. Nucleic Acids Res., 34, D108-110.</small>
</p>
