<h3 id="help_setting"><b>Setting page</b></h3>
<p>In the settings page you can change the parameters that influence the output. You can choose which tracks you want to visualize and add optional tracks.<br>
<b>NOTE</b> that some parameters are required and some are optional.<br><br>

Under <i>Visualization</i> you can select the level of details for the genome browser. You can choose between:</p>
<ul>
  <li><i>Gene</i> displays regions that include both protein-coding genes and non-coding RNA genes</li>
  <li><i>Transcript</i> displays transcripts and their stucture (exons, coding sequence) for both protein-coding genes and non-coding RNA genes</li>
</ul>

<br>
<h4 id="help_req"><b>Required parameters</b></h4>

<p>Under <i>peaks call format</i> you can select the consensus regulatory elements generated using the Narrow peaks dataset and/or the Broad peaks dataset as defined by the <a href="https://www.encodeproject.org" target="_blank">ENCODE</a> consortium.
</p>

<br>
<p>In <i>Consensus regulatory element</i> setting you can select which type of CRE (among: promoter, enhancer and active enhancer) you want to visualize</p>

<br>
<p>The <i>Consensus region tracks</i> setting permits you to visualize the consensus regolatory elements at different level of abstraction. You can choose between:</p>
<ul>
  <li><i>Global</i> displays the CREs generated combining all cell lines. It provides an overall landscape</li>
  <li><i>Tissue</i> displays the CREs generated combining the cell lines originated from the same tissue</li>
  <li><i>Cell line</i> displays the CREs generated combining replicates and experiments for the same cell line</li>
</ul>
<p><b>NOTE</b> that the lists of tissues and the tree of cell-lines are built based on the peak format selected. A different list of tissues and cell lines was available for Narrow and Broad peaks format. Then, selecting Narrow peak, Broad peak or both will provide a different list of tissue and cell line consensus regions. Thus, changing the peak format it will reset both tissue and cell line selections.
<br><br>
<figure><img src="./help/setting1.png" width="75%"><figcaption>Fig 2.1 - Setting page, required parameters</figcaption></figure>
</p>

<br>
<h4 id="help_tree"><b>Cell-line Tree selection</b></h4>
<p>Selecting the track <i>Cell line consensus regions</i>, a cell lines selection tree view will appear on the right of the webpage (Fig 2.2). The 198 cell lines are aggregated across the 38 tissue types. It is possible to check the box of a tissue in order to select all the cell lines of that specific tissue. Also, it is possible to expand the tissue and check the boxes of the cell lines of interest.
</p>
<figure><img src="./help/setting3.png" width="40%"><figcaption>Fig 2.2 - Setting page, cell line selection</figcaption></figure>

<br><br>
<h4 id="help_opt"><b>Optional parameters</b></h4>
<p>The <i>SNP track</i> displays all the single-nucleotide variants with a MAF greater or equal to 0.01 from <i>dbSNP v151</i></p>

<br>
<p>The <i>TSS track</i> displays the Transcription Start Sites with a confidence score grater or equal to 10. TSSs were determined by <i>SwitchGear Genomics</i></p>

<br>
<p>The <i>TBA information</i> display the Total Binding Affinity of a regulatory region for a specific TF.<br>NOTE that the TBA is available only for global and tissue consensus regions.</p>

<br>
<p>The <i>TBA p-value</i> reflects the significance of the TBA normalized score against a PFM-specific reference distribution of normalized TBA scores. Cutoff at 1e-05, which is set as default cutoff in CONREL, allows for stringent multiple-hypothesis correction at a specific CRE; more relaxed p-value cutoffs can be set for exploratory analyses.</p>

<br>
<p>The <i>Motifs count cutoff</i> puts a threshold on the number of aligned sequences used to built a TF motif. This is an option to limit the number of TF motifs in your output.<br>
A motif is usually derived from a set of aligned sequences. Lower count means lower sequences that supports a specific motif. This option excludes all the motifs with count (sequences aligned) below than threshold.
</p>

<br>
<p>The <i>Fraction motifs cutoff</i> puts a threshold on the fraction of regions with a significant TBA score for a specific motif. This is an option to limit the number of TF motifs in your output.<br>
This option excludes all the motifs that are enriched in TBA score in more regions (expressed as fraction over all the regions of the corresponding consensus) than the threshold provided.
</p><br><br>
<figure><img src="./help/setting2.png" width="75%"><figcaption>Fig 2.3 - Setting page, optional parameters</figcaption></figure>
